### PR TITLE
AF-2767: Kafka dataset provider

### PR DESCRIPTION
**Pull request by jesuino**
### Kafka Data Set Support for Dashbuilder

Ticket: https://issues.redhat.com/browse/AF-2767

This is a data set to retrieve real time metrics from Kafka. 

## Kafka Data sets

When creating data sets bear in mind that we need to define what kind of service we will be monitoring:

* BROKER: When using Broker you will be monitoring kafka server metrics, general metrics;
* CONSUMER: To monitor consumers and messages handled by a consumer;
* PRODUCER:  To monitor producers and messages created by a producer;

You can also filter the metrics you want to see using the Filter field. It filters by metrics name, type or domain. Attribute is not filtered.

The other fields are related to consumer and producer only, hence they are enabled only when producer/consumer is selected.

Client ID: The client Id - it is required for consumer/producer metrics;
Node ID: The node Id to monitor metrics for a specific node. When filled, topic is ignored
Topic: if you want metrics for a specific topic.

To test this dataset we must set up kafka with one node and a single topic.

### Kafka Setup

Kafka Setup, based on https://kafka-tutorials.confluent.io/kafka-console-consumer-producer-basics/kafka.html

1) Download Kafka
2) Start Zookeeper in a console

```
./bin/zookeeper-server-start.sh config/zookeeper.properties
```

3) In a new console start Kafka (Broker) - before export JMX_PORT=9999
```
export JMX_PORT=9999
./bin/kafka-server-start.sh config/server.properties
```

4) Create the topic using some console window
```
./bin/kafka-topics.sh --create --topic example-topic --bootstrap-server localhost:9092 --replication-factor 1 --partitions 1
```
5) In a new console start a Consumer - export JMX_PORT=9998
```
export JMX_PORT=9998
./bin/kafka-console-consumer.sh --group dashbuilder_test --topic example-topic --bootstrap-server localhost:9092
```
6) In a new console start the Producer
```
export JMX_PORT=9997
./bin/kafka-console-producer.sh --topic example-topic --broker-list localhost:9092
```
In the producer window you can type text that will be received by the consumer

image

### Data sets on Dashbuilder/BC

Now you can go to `BC -> Admin -> Data sets -> New Data Set` and you will notice a new type for Kafka:

image

Then you can create datasets for each parts of our Kafka setup using the following parameters:

Host: localhost
Port: 9999 for broker - 9998 for consumer and 9997 for producer

Client ID: console-producer for producer or consumer-dashbuilder_test-1 for consumer
Topic: example-topic

Examples:

BROKER:
image

BROKER OUTPUT:
image

CONSUMER:
image

CONSUMER OUTPUT
image

PRODUCER:
image

PRODUCER OUTPUT:
image

These 3 datasets can also be found in the following dashboard that can be imported into BC/Dashbuilder:

https://github.com/jesuino/dashbuilder-data/raw/master/dashboards/kafka_simple_dashboard.zip

You will basically also find three simple pages:

kafka_dashboard

Merge with:

https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1592
kiegroup/kie-soup#177
https://github.com/kiegroup/kie-wb-distributions/pull/1087

### CODE REVIEW

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
Hello @caponetto , @sthundat 

Would you please review this PR?

Thanks!

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by sthundat** [comment]
jenkins execute fdb

**Review by caponetto — COMMENTED** [review]
Hi @jesuino, thanks for the PR! Very interesting stuff.
The code looks great to me. I've only left a few minor comments.
I still want to give it a try, can you share a war file for BC?
And I'll review the `kie-soup` PR as soon as possible.

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
Hello @caponetto 

I made the requested changes, thanks!

I am building a local business-central-webapp because Jenkins FDB is failing to build. It may be due a test failure from this PR, so I will fire another FDB, but in any case I will share a WAR tomorrow.

Thanks!

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
Here's a link to a Business Central Web App that I built locally. Deploy it on Wildfly 19.0.1 - a suggestion is use the jbpm server distribution (7.50.0.FInal) and just replace business-central.war 

https://drive.google.com/file/d/1XHQGowI_-1aYAl9QjUeSS7uvMGPV5Q5Q/view?usp=sharing

**Comment by caponetto** [comment]
Hi @jesuino, here are two observations after doing some tests:

1. Should this `size must be between 4 and ...` validation for `port` be there?
tooltip

2. If the connection can't be established, can we have a more friendly message? (e.g. in this case I've put a wrong port)
popup

Other than this, pretty nice stuff. :)

**Comment by jesuino** [comment]
Hello @caponetto 

Thank you so much for verifying the WAR. 


1) I changed the validation to be more close to a port validation. Now noticed that the message is not correctly encoded - this is not related to this PR, but simply how the hint is built by datasets editors - I can revisit later.

image

2) I did changed the message that is thrown from the Provider, but this is also something at the editors level: it, for some reason, shows the root cause message, which prevents me from adding a human friendly message. This behaviour can also be changed.

The reason I don't do all these improvements is because we were discussing about a new data set editor perspective, which would be a new view created from scratch. But we may have a bulk of improvements for this old data sets editors when we have some time to work on it!

**Comment by caponetto** [comment]
Great, @jesuino! It makes sense to me. :)

**Comment by jesuino** [comment]
I changed the message to avoid the encoding issue:

image

The commit is in kie-soup: https://github.com/kiegroup/kie-soup/pull/177/commits/6ef072bfca64db4f3968676d4be7327c028d4501

File: dashbuilder/dashbuilder-runtime/src/main/webapp/WEB-INF/jboss-deployment-structure.xml [hunk 1]
```diff
@@ -36,6 +36,8 @@
 			<module name="org.jboss.resteasy.resteasy-multipart-provider" />
 			<module name="org.apache.xerces" />
 			<module name="org.apache.xalan" />
+		    <!-- For Kafka Connector -->
+            <module name="org.jboss.remoting-jmx"/>
```
Comment (caponetto) [inline dashbuilder/dashbuilder-runtime/src/main/webapp/WEB-INF/jboss-deployment-structure.xml:40]: One last thing: this indentation should be aligned with line 38. 😛

**Comment by sthundat** [comment]
Hello @jesuino ,
Thanks for the PR.
Below are  some of the observation on kafka dataset,
1.if I select wrong port (port of consumer/producer)for broker not getting connection refuse exception in UI.if i select some other values(ex:9992) connection refused exception is throwing as expected.
brokerwrongport
wrong port selection of broker result in below page
brokerwrongportdetailswithnodata
2.Is it possible to remove target field overlay information that is displaying just above the drop down.This sometimes preventing users to select the target value from the dropdown.
please note that these are my observation as per current testing.will continue testing next day also.
Thanks,
Savitha

**Review by caponetto — APPROVED** [review]

**Comment by jesuino** [comment]
Hello @sthundat,

Thanks for your review! Here are my answers for your inquires:

1) That's unfortunately expected. The way the data set works is by scanning MBeans for a specific target: if you select Broker it will scan all Broker MBeans in the provided server. If there's no MBean then it returns an empty data set. It will throw exception only if the server is not available!

2) Could you please share a screenshot or guide me to reproduce this behavior? I know that there's a bug with editors that unfortunately do not select the values in drop downs (it happens also with Remote Server), but it would not be related to this specific provider.

Looking forward for your next reviews, thanks!

**Comment by sthundat** [comment]
Hi @jesuino ,
Thanks for your inputs.
Regarding second point,Please find the video below.As you said this is an issue with other dataset as well.
dropdownissue
Also if client id field (producer/consumer) is empty user is prompted with exception on screen.But providing wrong client id/topic result in empty dataset.Is it possible to validate this and gives user some information on  what is wrong on input?
Thanks,
Savitha

**Comment by jesuino** [comment]
Hello @sthundat 

I made the requested changes:

1) Changed the tooltip to show on top instead on bottom

image

2) Added an error for when there's no metric (the change is in PR for kie soup https://github.com/kiegroup/kie-soup/pull/177/commits/17b4635e0988ce9d8fc4f522442901144eb8610a):

image

I also changed the root cause of exception in dataset provider, hiding the root cause from the exception threw by the provider, but logging the root cause in DEBUG mode)

image

Finally I changed validation for host, it may contain less characters

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by sthundat** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
Hello,

I built a WAR with the latest changes and made it available here

https://drive.google.com/file/d/1XHQGowI_-1aYAl9QjUeSS7uvMGPV5Q5Q/view?usp=sharing

Thanks!

**Comment by sthundat** [comment]
Hi @jesuino ,
Thanks for the changes.Now the validation looks good with the new error messages.
Approving the PR.
Thanks,
Savitha

**Review by sthundat — APPROVED** [review]

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
There are multiple test failures, but none of them are related to Dashbuilder

https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/fdb/job/appformer-master.fdb/136/#showFailuresLink

Will fire a new build to see if the errors were corrected

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
The failed test is unrelated to this PR


image

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
Business Central was finally built with this PR - I downloaded and test locally! IMO we are safe to merge. The failing test is unrelated

image

**Review by ederign — APPROVED** [review]
LGTM!

**Pull request merged**